Mathlib/Data/String/Basic.lean

```diff
@@ -1,200 +0,0 @@
-@[simp]
-theorem lt_iff_toList_lt : ∀ {s₁ s₂ : String}, s₁ < s₂ ↔ s₁.toList < s₂.toList
-  | ⟨s₁⟩, ⟨s₂⟩ => show ltb ⟨⟨s₁⟩, 0⟩ ⟨⟨s₂⟩, 0⟩ ↔ s₁ < s₂ by
-    induction s₁ generalizing s₂ <;> cases s₂
-    · unfold ltb; decide
-    · rename_i c₂ cs₂; apply iff_of_true
-      · unfold ltb
-        #adaptation_note /-- v4.7.0-rc1 exclude reduceMk from simp -/
-        simp [-reduceMk, Iterator.hasNext, csize_pos]
-      · apply List.nil_lt_cons
-    · rename_i c₁ cs₁ ih; apply iff_of_false
-      · unfold ltb
-        #adaptation_note /-- v4.7.0-rc1 exclude reduceMk from simp -/
-        simp [-reduceMk, Iterator.hasNext]
-      · apply not_lt_of_lt; apply List.nil_lt_cons
-    · rename_i c₁ cs₁ ih c₂ cs₂; unfold ltb
-      simp only [Iterator.hasNext, Pos.byteIdx_zero, endPos, utf8ByteSize, utf8ByteSize.go,
-        add_pos_iff, csize_pos, or_true, decide_eq_true_eq, ↓reduceIte, Iterator.curr, get,
-        utf8GetAux, Iterator.next, next, Bool.ite_eq_true_distrib]
-      split_ifs with h
-      · subst c₂
-        suffices ltb ⟨⟨c₁ :: cs₁⟩, (0 : Pos) + c₁⟩ ⟨⟨c₁ :: cs₂⟩, (0 : Pos) + c₁⟩ =
-          ltb ⟨⟨cs₁⟩, 0⟩ ⟨⟨cs₂⟩, 0⟩ by rw [this]; exact (ih cs₂).trans List.Lex.cons_iff.symm
-        apply ltb_cons_addChar
-      · refine ⟨List.Lex.rel, fun e ↦ ?_⟩
-        cases e <;> rename_i h'
-        · contradiction
-        · assumption
-#align string.lt_iff_to_list_lt String.lt_iff_toList_lt
-
-instance LE : LE String :=
-  ⟨fun s₁ s₂ ↦ ¬s₂ < s₁⟩
-#align string.has_le String.LE
-
-instance decidableLE : @DecidableRel String (· ≤ ·) := by
-  simp only [LE]
-  infer_instance -- short-circuit type class inference
-#align string.decidable_le String.decidableLE
-
-@[simp]
-theorem le_iff_toList_le {s₁ s₂ : String} : s₁ ≤ s₂ ↔ s₁.toList ≤ s₂.toList :=
-  (not_congr lt_iff_toList_lt).trans not_lt
-#align string.le_iff_to_list_le String.le_iff_toList_le
-
-theorem toList_inj {s₁ s₂ : String} : s₁.toList = s₂.toList ↔ s₁ = s₂ :=
-  ⟨congr_arg mk, congr_arg toList⟩
-#align string.to_list_inj String.toList_inj
-
-theorem nil_asString_eq_empty : [].asString = "" :=
-  rfl
-#align string.nil_as_string_eq_empty String.nil_asString_eq_empty
-
-@[simp]
-theorem toList_empty : "".toList = [] :=
-  rfl
-#align string.to_list_empty String.toList_empty
-
-theorem asString_inv_toList (s : String) : s.toList.asString = s :=
-  rfl
-#align string.as_string_inv_to_list String.asString_inv_toList
-
-#align string.to_list_singleton String.data_singleton
-
-theorem toList_nonempty : ∀ {s : String}, s ≠ "" → s.toList = s.head :: (s.drop 1).toList
-  | ⟨s⟩, h => by
-    cases s with
-    | nil => simp at h
-    | cons c cs =>
-      simp only [toList, data_drop, List.drop_succ_cons, List.drop_zero, List.cons.injEq, and_true]
-      rfl
-#align string.to_list_nonempty String.toList_nonempty
-
-@[simp]
-theorem head_empty : "".data.head! = default :=
-  rfl
-#align string.head_empty String.head_empty
-
-#align string.popn_empty String.drop_empty
-
-instance : LinearOrder String where
```

Comment:
I made this remark privately, but for transparency I'll make it here too.

There generic approach of mathlib means that many `instance`s are never actually used; but that doesn't mean that the instance is useless. In this case, the instance enables things like
```lean
import Mathlib
#eval ["hello", "world"].mergeSort (· ≤ ·)
```
work. Removing this seems like it is going to make it needlessly harder to write scripts and meta code.